File: phantom-mac-app/Sources/Phantom/SpeechTranscriptionClient.swift
import Foundation

@MainActor
final class SpeechTranscriptionClient {
    private let backend: BackendClient
    private let rotation: APIRotationManager

    init(backend: BackendClient, rotation: APIRotationManager) {
        self.backend = backend
        self.rotation = rotation
    }

    static func selfCheck() -> Bool {
        let silent = Data(repeating: 0, count: 16_000)
        let voiced = Data(repeating: 0x10, count: 16_000)
        let wave = wav(silent)
        return !containsSpeech(silent)
            && containsSpeech(voiced)
            && String(data: wave.prefix(4), encoding: .ascii) == "RIFF"
            && wave.count == silent.count + 44
    }

    func transcribe(
        pcm16: Data,
        session: AuthSession,
        managed: Bool,
        provider: String,
        model: String,
        language: String,
        useChatKeys: Bool
    ) async throws -> String {
        guard Self.containsSpeech(pcm16) else { return "" }
        let wav = Self.wav(pcm16)
        if managed { return try await backend.transcribeManagedSpeech(accessToken: session.accessToken, wav: wav, language: language) }
        guard !model.isEmpty else { throw BackendError.server("Choose a speech model in Settings.") }
        let keys = useChatKeys ? rotation.keys(for: provider) : rotation.speechKeys(provider: provider)
        guard !keys.isEmpty else { throw BackendError.server("Add a speech API key or enable shared chat keys.") }

        var lastError: Error?
        for attempt in 0..<min(2, keys.count) {
            let scope = "\(provider):\(useChatKeys ? "chat" : "dedicated")"
            guard let selected = nextKey(provider: scope, keys: keys) else { break }
            do {
                let text = try await Self.post(wav: wav, provider: provider, model: model, language: language, apiKey: selected.value)
                clearCooldown(provider: scope, index: selected.index)
                return text
            } catch {
                lastError = error
                let decision = rotation.classify(error)
                if decision.canRotateCredential { setCooldown(provider: scope, index: selected.index, duration: decision.cooldown) }
                if !ProviderResiliencePolicy.canRetry(decision, attempt: attempt + 1, maxAttempts: 2, hasOutput: false) { break }
            }
        }
        throw lastError ?? BackendError.server("No speech API key is currently available.")
    }

    private func nextKey(provider: String, keys: [String]) -> (index: Int, value: String)? {
        let defaults = UserDefaults.standard
        let indexKey = "speech.rotation.\(provider.lowercased()).lastKey"
        let last = defaults.object(forKey: indexKey) == nil ? -1 : defaults.integer(forKey: indexKey)
        for offset in 1...keys.count {
            let index = (last + offset) % keys.count
            if cooldowns(provider)[String(index), default: 0] > Date().timeIntervalSince1970 { continue }
            defaults.set(index, forKey: indexKey)
            return (index, keys[index])
        }
        return nil
    }

    private func cooldowns(_ provider: String) -> [String: Double] {
        UserDefaults.standard.dictionary(forKey: "speech.rotation.\(provider.lowercased()).cooldowns") as? [String: Double] ?? [:]
    }

    private func setCooldown(provider: String, index: Int, duration: TimeInterval) {
        var values = cooldowns(provider)
        values[String(index)] = Date().addingTimeInterval(duration).timeIntervalSince1970
        UserDefaults.standard.set(values, forKey: "speech.rotation.\(provider.lowercased()).cooldowns")
    }

    private func clearCooldown(provider: String, index: Int) {
        var values = cooldowns(provider)
        values.removeValue(forKey: String(index))
        UserDefaults.standard.set(values, forKey: "speech.rotation.\(provider.lowercased()).cooldowns")
    }

    private static func post(wav: Data, provider: String, model: String, language: String, apiKey: String) async throws -> String {
        let urls = [
            "ChatGPT": "https://api.openai.com/v1/audio/transcriptions",
            "Groq": "https://api.groq.com/openai/v1/audio/transcriptions",
            "Mistral": "https://api.mistral.ai/v1/audio/transcriptions"
        ]
        guard let endpoint = urls[provider], let url = URL(string: endpoint) else { throw BackendError.server("Unsupported speech provider.") }
        let boundary = "Phantom-\(UUID().uuidString)"
        var body = Data()
        func field(_ name: String, _ value: String) { body.append(Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n".utf8)) }
        field("model", model); field("language", language); field("response_format", "json")
        body.append(Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"speech.wav\"\r\nContent-Type: audio/wav\r\n\r\n".utf8))
        body.append(wav); body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        var request = URLRequest(url: url)
        request.httpMethod = "POST"; request.timeoutInterval = 60; request.httpBody = body
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else { throw BackendError.http(status, "Speech provider failed (\(status)).") }
        return (try JSONDecoder().decode(SpeechText.self, from: data)).text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func wav(_ pcm: Data) -> Data {
        var result = Data()
        result.append(Data("RIFF".utf8)); result.appendLE(UInt32(36 + pcm.count)); result.append(Data("WAVEfmt ".utf8))
        result.appendLE(UInt32(16)); result.appendLE(UInt16(1)); result.appendLE(UInt16(1)); result.appendLE(UInt32(16_000))
        result.appendLE(UInt32(32_000)); result.appendLE(UInt16(2)); result.appendLE(UInt16(16)); result.append(Data("data".utf8)); result.appendLE(UInt32(pcm.count)); result.append(pcm)
        return result
    }

    private static func containsSpeech(_ pcm: Data) -> Bool {
        let values = pcm.withUnsafeBytes { Array($0.bindMemory(to: Int16.self)) }
        guard !values.isEmpty else { return false }
        var total: Int64 = 0
        var count = 0
        for index in stride(from: 0, to: values.count, by: 8) { total += Int64(abs(Int(values[index]))); count += 1 }
        return count > 0 && total / Int64(count) >= 120
    }
}

private struct SpeechText: Decodable { let text: String }

private extension Data {
    mutating func appendLE<T: FixedWidthInteger>(_ value: T) {
        var little = value.littleEndian
        Swift.withUnsafeBytes(of: &little) { append(contentsOf: $0) }
    }
}

File: phantom-windows-app/Services/SpeechTranscriptionClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using SecureOverlay.Infrastructure.Hosted.Contracts;

namespace SecureOverlay.Services
{
    public sealed class SpeechTranscriptionClient
    {
        private static readonly HttpClient HttpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(60) };
        private readonly AppSettings _settings;
        private readonly string _backendBaseUrl;
        private readonly string _accessToken;
        private readonly bool _managed;

        public SpeechTranscriptionClient(AppSettings settings, string backendBaseUrl, string accessToken, bool managed)
        {
            _settings = settings;
            _backendBaseUrl = backendBaseUrl.TrimEnd('/');
            _accessToken = accessToken;
            _managed = managed;
        }

        public async Task<string> TranscribePcm16Async(byte[] pcm16, CancellationToken cancellationToken)
        {
            if (!ContainsSpeech(pcm16))
            {
                Log.WriteLine("Cloud speech skipped reason=no_speech_detected");
                return string.Empty;
            }
            var wav = BuildWav(pcm16, 16_000, 1);
            if (_managed) return await TranscribeManagedAsync(wav, cancellationToken);

            var keys = GetByoKeys();
            if (keys.Count == 0) throw new InvalidOperationException("No speech API key is configured.");
            Exception? lastError = null;
            for (var attempt = 0; attempt < Math.Min(keys.Count, ProviderResiliencePolicy.ByoDesktopMaxAttempts); attempt++)
            {
                var selected = NextAvailableKey(keys);
                if (selected == null) break;
                try
                {
                    var text = await PostTranscriptionAsync(GetProviderUrl(_settings.SpeechProviderId), selected.Value.Key, wav, _settings.SpeechModelId, _settings.SpeechLanguage, cancellationToken);
                    _settings.SpeechRotationState.KeyCooldownUntilUtc.Remove(CooldownName(selected.Value.Index));
                    SettingsManager.Save(_settings);
                    Log.WriteLine($"BYO speech provider completed provider={_settings.SpeechProviderId} model={_settings.SpeechModelId} key_slot={selected.Value.Index + 1} attempt={attempt + 1}");
                    return text;
                }
                catch (Exception ex) when (ex is not OperationCanceledException)
                {
                    lastError = ex;
                    var failure = ProviderResiliencePolicy.Classify(ex.Message);
                    Log.WriteLine($"BYO speech provider failed provider={_settings.SpeechProviderId} model={_settings.SpeechModelId} attempt={attempt + 1} error_code={failure.ErrorCode}");
                    if (failure.CanRotateCredential)
                    {
                        _settings.SpeechRotationState.KeyCooldownUntilUtc[CooldownName(selected.Value.Index)] = DateTime.UtcNow.Add(failure.Cooldown);
                        SettingsManager.Save(_settings);
                    }
                    if (!ProviderResiliencePolicy.CanRetry(failure, attempt + 1, ProviderResiliencePolicy.ByoDesktopMaxAttempts, false)) break;
                }
            }
            throw lastError ?? new InvalidOperationException("No speech API key is currently available.");
        }

        private async Task<string> TranscribeManagedAsync(byte[] wav, CancellationToken cancellationToken)
        {
            using var form = BuildForm(wav, model: null, _settings.SpeechLanguage);
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{_backendBaseUrl}/api/desktop/speech/transcribe") { Content = form };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
            using var response = await HttpClient.SendAsync(request, cancellationToken);
            var body = await response.Content.ReadAsStringAsync(cancellationToken);
            if (!response.IsSuccessStatusCode) throw new HttpRequestException($"Managed speech failed ({(int)response.StatusCode}).");
            var result = JsonConvert.DeserializeObject<SpeechTranscriptionResponseDto>(body);
            Log.WriteLine($"Managed speech provider completed provider={result?.ProviderId ?? "unknown"} model={result?.ModelId ?? "unknown"}");
            return result?.Text?.Trim() ?? string.Empty;
        }

        private static async Task<string> PostTranscriptionAsync(string url, string apiKey, byte[] wav, string model, string language, CancellationToken cancellationToken)
        {
            using var form = BuildForm(wav, model, language);
            using var request = new HttpRequestMessage(HttpMethod.Post, url) { Content = form };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
            using var response = await HttpClient.SendAsync(request, cancellationToken);
            var body = await response.Content.ReadAsStringAsync(cancellationToken);
            if (!response.IsSuccessStatusCode) throw new HttpRequestException($"Speech provider failed ({(int)response.StatusCode}).");
            using var reader = new Newtonsoft.Json.JsonTextReader(new StringReader(body));
            var result = Newtonsoft.Json.Linq.JObject.Load(reader);
            return result.Value<string>("text")?.Trim() ?? string.Empty;
        }

        private static MultipartFormDataContent BuildForm(byte[] wav, string? model, string language)
        {
            var form = new MultipartFormDataContent();
            var audio = new ByteArrayContent(wav);
            audio.Headers.ContentType = new MediaTypeHeaderValue("audio/wav");
            form.Add(audio, string.IsNullOrWhiteSpace(model) ? "audio" : "file", "speech.wav");
            if (!string.IsNullOrWhiteSpace(model))
            {
                form.Add(new StringContent(model), "model");
                form.Add(new StringContent("json"), "response_format");
            }
            if (!string.IsNullOrWhiteSpace(language)) form.Add(new StringContent(language), "language");
            return form;
        }

        private List<string> GetByoKeys()
        {
            if (!_settings.UseChatProviderApiKeysForSpeech)
                return _settings.SpeechApiKeys.TryGetValue(_settings.SpeechProviderId, out var dedicated) ? dedicated : new List<string>();
            return _settings.SpeechProviderId switch
            {
                "ChatGPT" => _settings.ChatGPTApiKeys,
                "Groq" => _settings.GroqApiKeys,
                "Mistral" => _settings.MistralApiKeys,
                _ => new List<string>()
            };
        }

        private (int Index, string Key)? NextAvailableKey(IReadOnlyList<string> keys)
        {
            var state = _settings.SpeechRotationState;
            var scope = RotationScope;
            var last = state.LastKeyIndex.TryGetValue(scope, out var value) ? value : -1;
            for (var offset = 1; offset <= keys.Count; offset++)
            {
                var index = (last + offset) % keys.Count;
                if (state.KeyCooldownUntilUtc.TryGetValue(CooldownName(index), out var until) && until > DateTime.UtcNow) continue;
                state.LastKeyIndex[scope] = index;
                return (index, keys[index]);
            }
            return null;
        }

        private string RotationScope => $"{_settings.SpeechProviderId}:{(_settings.UseChatProviderApiKeysForSpeech ? "chat" : "dedicated")}";
        private string CooldownName(int index) => $"{RotationScope}:{index}";

        private static string GetProviderUrl(string provider) => provider switch
        {
            "ChatGPT" => "https://api.openai.com/v1/audio/transcriptions",
            "Groq" => "https://api.groq.com/openai/v1/audio/transcriptions",
            "Mistral" => "https://api.mistral.ai/v1/audio/transcriptions",
            _ => throw new InvalidOperationException("Unsupported speech provider.")
        };

        private static byte[] BuildWav(byte[] pcm, int sampleRate, short channels)
        {
            using var output = new MemoryStream(44 + pcm.Length);
            using var writer = new BinaryWriter(output, Encoding.ASCII, true);
            writer.Write(Encoding.ASCII.GetBytes("RIFF")); writer.Write(36 + pcm.Length); writer.Write(Encoding.ASCII.GetBytes("WAVE"));
            writer.Write(Encoding.ASCII.GetBytes("fmt ")); writer.Write(16); writer.Write((short)1); writer.Write(channels);
            writer.Write(sampleRate); writer.Write(sampleRate * channels * 2); writer.Write((short)(channels * 2)); writer.Write((short)16);
            writer.Write(Encoding.ASCII.GetBytes("data")); writer.Write(pcm.Length); writer.Write(pcm);
            return output.ToArray();
        }

        private static bool ContainsSpeech(byte[] pcm)
        {
            if (pcm.Length < 2) return false;
            long total = 0;
            var count = 0;
            for (var index = 0; index + 1 < pcm.Length; index += 16)
            {
                total += Math.Abs((int)BitConverter.ToInt16(pcm, index));
                count++;
            }
            return count > 0 && total / count >= 120;
        }
    }
}
